File: ts/cola/src/cola-secuencial.ts
export class ColaSecuencial {
  private items: number[];
  private dimension: number;
  private cantidad: number;
  private ultimo: number;
  private primero: number;

  constructor(dimension: number) {
    this.dimension = dimension;
    this.cantidad = 0;
    this.ultimo = 0;
    this.primero = 0;
    this.items = [];
  }

  insertar(item: number): number {
    if (this.cantidad == this.dimension) {
      console.log("Cola llena.");
      return -1;
    }

    this.items[this.ultimo] = item;
    this.ultimo = (this.ultimo + 1) % this.dimension;
    this.cantidad++;
    return item;
  }

  suprimir(): number {
    if (this.vacia()) {
      console.log("Cola vacia.");
      return -1;
    }
    let eliminado = this.items[this.primero];
    this.primero = (this.primero + 1) % this.dimension;
    return eliminado;
  }

  recorrer() {
    if (this.vacia()) {
      console.log("Cola vacia.");
      return;
    }
    let cola: number[] = [];
    let i = this.primero;
    for (let j = this.ultimo; i > -1; i--) {
      cola.push(this.items[i]);
    }
    console.log(cola);
  }

  vacia(): boolean {
    return this.cantidad == 0;
  }
}

File: estructuras/ColaSecuencial.py
import numpy as np


class ColaSecuencial:
    __dimension: int
    __cantidad: int
    __ultimo: int
    __primero: int
    __items: np.ndarray

    def __init__(self, dimension):
        self.__dimension = dimension
        self.__cantidad = 0
        self.__ultimo = 0
        self.__primero = 0
        self.__items = np.empty(dimension, dtype=int)

    def insertar(self, item):
        if self.__cantidad == self.__dimension:
            print("Cola llena.")
            return None
        self.__items[self.__ultimo] = item
        self.__ultimo = (self.__ultimo + 1) % self.__dimension
        self.__cantidad += 1
        return item

    def suprimir(self):
        if self.vacia():
            print("Cola vacia.")
            return None
        eliminado = self.__items[self.__primero]
        self.__primero = (self.__primero + 1) % self.__dimension
        self.__cantidad -= 1
        return eliminado

    def recorrer(self):
        if self.vacia():
            print("Cola vacia.")
            return None
        i = self.__primero
        for _ in range(self.__cantidad):
            print(self.__items[i], end=" ")
            i = (i + 1) % self.__dimension
        print("\n")

    def vacia(self):
        return self.__cantidad == 0


if __name__ == "__main__":
    cola = ColaSecuencial(4)

    cola.insertar(1)
    cola.insertar(2)
    cola.insertar(3)
    cola.insertar(4)
    cola.insertar(5)

    cola.recorrer()

    cola.suprimir()
    cola.suprimir()

    cola.recorrer()
